### PR TITLE
[fix] Fucking ugly workaround for the goddamn dependency nighmare from sury djeezus kraiste

### PR DESCRIPTION
**Pull request by alexAubin**
## The problem

c.f. [this hell](https://github.com/YunoHost/issues/issues/1407)

## Solution

Lol just readd the sury repo in the middle of the `install_app_dependency` helper if we detect that sury fucked everything and we need other php deps

## PR Status

Yolo commited without any test, should be tested ...

## How to test

Install pixelfed then try to install e.g. nextcloud

## Validation

- [ ] Principle agreement 0/2 : 
- [ ] Quick review 0/1 : 
- [ ] Simple test 0/1 : 
- [ ] Deep review 0/1 :

### CODE REVIEW

**Comment by decentral1se** [comment]
OK, thanks for facing this one down. I've gone ahead and installed pixelfed and then nextcloud on a fresh VPS with this patch applied. During the Nextcloud installation, I got the following logs:

```
54605 DEBUG + echo '[###########++++.....] > Installing nextcloud...'
54605 DEBUG + set -x
54605 DEBUG + chown -R nextcloud: /var/www/nextcloud /home/yunohost.app/nextcloud/data
54806 DEBUG ++ ynh_string_random --length=6
54806 DEBUG ++ local legacy_args=l
54807 DEBUG ++ args_array=([l]=length=)
54807 DEBUG ++ declare -Ar args_array
54807 DEBUG ++ local length
54807 DEBUG ++ ynh_handle_getopts_args --length=6
54807 DEBUG ++ set +x
54807 DEBUG ++ length=6
54807 DEBUG ++ sed -n 's/\(.\{6\}\).*/\1/p'
54808 DEBUG ++ tr -c -d A-Za-z0-9
54808 DEBUG ++ dd if=/dev/urandom bs=1 count=1000
54808 DEBUG + exec_occ maintenance:install --database mysql --database-name nextcloud --database-user nextcloud --database-pass RqYJ8wbZ6iKxjbwczaB24ypS --admin-user admin --admin-pass SohYkk --data-dir /home/yunohost.app/nextcloud/data
54808 DEBUG + cd /var/www/nextcloud
54808 DEBUG + exec_as nextcloud php occ --no-interaction --no-ansi maintenance:install --database mysql --database-name nextcloud --database-user nextcloud --database-pass RqYJ8wbZ6iKxjbwczaB24ypS --admin-user admin --admin-pass SohYkk --data-dir /home/yunohost.app/nextcloud/data
54808 DEBUG + local USER=nextcloud
54809 DEBUG + shift 1
54809 DEBUG ++ whoami
54809 DEBUG + [[ nextcloud = root ]]
54809 DEBUG + sudo -u nextcloud php occ --no-interaction --no-ansi maintenance:install --database mysql --database-name nextcloud --database-user nextcloud --database-pass RqYJ8wbZ6iKxjbwczaB24ypS --admin-user admin --admin-pass SohYkk --data-dir /home/yunohost.app/nextcloud/data
55010 DEBUG PHP module zip not installed.
55011 WARNING Unable to install Nextcloud
55012 DEBUG Please ask your server administrator to install the module.
55012 DEBUG 
55012 DEBUG An unhandled exception has been thrown:
55013 DEBUG Exception: Environment not properly prepared. in /var/www/nextcloud/lib/private/Console/Application.php:166
55013 DEBUG Stack trace:
55014 DEBUG #0 /var/www/nextcloud/console.php(95): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
55014 DEBUG #1 /var/www/nextcloud/occ(11): require_once('/var/www/nextcl...')
55015 DEBUG #2 {main}+ ynh_die '--message=Unable to install Nextcloud'
55016 DEBUG + local legacy_args=mc
55016 DEBUG + args_array=([m]=message= [c]=ret_code=)
55017 DEBUG + declare -Ar args_array
55017 DEBUG + local message
55018 DEBUG + local ret_code
55018 DEBUG + ynh_handle_getopts_args '--message=Unable to install Nextcloud'
55019 DEBUG + set +x
55020 DEBUG + echo 'Unable to install Nextcloud'
55020 DEBUG + exit 1
55021 DEBUG + ynh_exit_properly
55021 DEBUG + local exit_code=1
55022 DEBUG + '[' 1 -eq 0 ']'
55022 DEBUG + trap '' EXIT
55023 DEBUG + set +eu
55023 DEBUG + sleep 0.5
55526 DEBUG + ynh_print_err '--message=!!\n  nextcloud'\''s script has encountered an error. Its execution was cancelled.\n!!'
55527 WARNING [ERR] !!
55527 DEBUG + local legacy_args=m
55527 WARNING   nextcloud's script has encountered an error. Its execution was cancelled.
55528 DEBUG + args_array=([m]=message=)
55528 WARNING !!
55528 DEBUG + declare -Ar args_array
55528 WARNING Please find here an extract of the log before the crash:
55528 DEBUG + local message
55529 WARNING [DEBUG]: DEBUG   - + local USER=nextcloud
55529 DEBUG + ynh_handle_getopts_args '--message=!!\n  nextcloud'\''s script has encountered an error. Its execution was cancelled.\n!!'
55529 WARNING [DEBUG]: DEBUG   - + shift 1
55529 DEBUG + set +x
55529 WARNING [DEBUG]: DEBUG   - ++ whoami
55530 DEBUG + ynh_print_log '\e[91m\e[1m[ERR]\e[0m !!\n  nextcloud'\''s script has encountered an error. Its execution was cancelled.\n!!'
55530 WARNING [DEBUG]: DEBUG   -
55530 DEBUG + echo -e '\e[91m\e[1m[ERR]\e[0m !!\n  nextcloud'\''s script has encountered an error. Its execution was cancelled.\n!!'
55531 WARNING [DEBUG]: DEBUG   - + sudo -u nextcloud php occ --no-interaction --no-ansi maintenance:install --database mysql --database-name nextcloud --database-user nextcloud --database-pass RqYJ8wbZ6iKxjbwczaB24ypS --admin-user admin --admin-pass SohYkk --data-dir /home/yunohost.app/nextcloud/data
55531 DEBUG + '[' cli == cli ']'
55532 WARNING [DEBUG]: DEBUG   - PHP module zip not installed.
```

And actually, I do not see the repository configured in my sources folder:

```
root@ynhdev:~# ls -lha /etc/apt/sources.list.d/
total 20K
drwxr-xr-x 2 root root 4.0K Sep 25 15:18 .
drwxr-xr-x 6 root root 4.0K Sep 24 17:11 ..
-rw-r--r-- 1 root root  243 Jul 14 06:07 hetzner-mirror.list
-rw-r--r-- 1 root root   83 Jul 14 06:07 hetzner-security-updates.list
-rw-r--r-- 1 root root   53 Sep 24 12:04 yunohost.list
```

And then when I tried to manually install this missing package I got:

```
root@ynhdev:~# apt install php7.0-zip
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 php7.0-zip : Depends: php7.0-common (= 7.0.33-0+deb9u5) but 7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e is to be installed
E: Unable to correct problems, you have held broken packages.
```

However, once I manually added the repository, I could install nextcloud!

So, I think the conditional check just isn't being triggered ... I'll review the patch again.

File: data/helpers.d/apt [hunk 1]
```diff
@@ -218,6 +218,27 @@ ynh_install_app_dependencies () {
...
+    # https://github.com/YunoHost/issues/issues/1407
+    # 
+    # If we require to install php dependency
+    if echo $dependencies | grep -q 'php';
+    then
+        # And we have packages from sury installed (7.0.33-10+weirdshiftafter instead of 7.0.33+1 on debian)
+        if dpkg --list | grep php | grep -q "7.0.33-10"
+        then
+            # And sury ain't already installed
+            if ! grep -nrq "sury" /etc/apt/sources.list*
```
Comment (decentral1se) [inline data/helpers.d/apt]: Ah, we need to look inside `/etc/apt/sources.list.d/`, no?

Comment (alexAubin) [reply data/helpers.d/apt]: Hmmmmyup but it should be caught in the `/etc/apt/sources.list*` (note the wildcard) ?

Comment (decentral1se) [reply data/helpers.d/apt]: Ah, I'm re-installing both apps again and I'll test this by hand once more. Two secs.

Comment (decentral1se) [reply data/helpers.d/apt]:
```
root@ynhdev:~# dpkg --list | grep php | grep 7.0.33
ii  php7.0-cli                            7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        command-line interpreter for the PHP scripting language
ii  php7.0-common                         7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        documentation, examples and common module for PHP
ii  php7.0-curl                           7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        CURL module for PHP
ii  php7.0-fpm                            7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        server-side, HTML-embedded scripting language (FPM-CGI binary)
ii  php7.0-gd                             7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        GD module for PHP
ii  php7.0-intl                           7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        Internationalisation module for PHP
ii  php7.0-json                           7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        JSON module for PHP
ii  php7.0-ldap                           7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        LDAP module for PHP
ii  php7.0-mbstring                       7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        MBSTRING module for PHP
ii  php7.0-mcrypt                         7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        libmcrypt module for PHP
ii  php7.0-mysql                          7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        MySQL module for PHP
ii  php7.0-opcache                        7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        Zend OpCache module for PHP
ii  php7.0-readline                       7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        readline module for PHP
ii  php7.0-xml                            7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        DOM, SimpleXML, WDDX, XML, and XSL module for PHP
ii  php7.0-zip                            7.0.33-11+0~20190923.20+debian9~1.gbpd05c7e       amd64        Zip module for PHP
```

Where "-10" doesn't make the correct match ... :sweat:

Comment (alexAubin) [reply data/helpers.d/apt]: Hmwelll, I guess we need to either use a real version comparison tool to be more robust, or grep somethind else like the "+debian9~" part of the "gbp" thing (the hexadecimal stuff seems to be changing over time) or idk :/

Comment (alexAubin) [reply data/helpers.d/apt]: Oooooor `grep -v "7.0.33-0+deb9u5"` which is debian's version

Comment (decentral1se) [reply data/helpers.d/apt]: `dpkg --list | grep php | grep -v "7.0.33-0+deb9u5` does return what we need!

Comment (decentral1se) [reply data/helpers.d/apt]: Yes, I've patched that and I can isntall Nextcloud successfully. It should be fine :hand:

**Pull request merged**